### ISSUE TITLE
Issue connecting when using nxos_acl

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#29473]
From @RNanney on 2016-11-18T22:39:18Z

##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

nxos_acl

##### ANSIBLE VERSION

```
ansible 2.2.0.0
  config file = /home/robert.nanney/ansible.cfg
  configured module search path = Default w/o overrides

```

##### SUMMARY
Can not connect with transport: cli to the device

##### STEPS TO REPRODUCE

```
---
- hosts: all
  connection: local
  gather_facts: false
  vars:
    cli:
      transport: cli

  tasks:
    - name: show hostname
      nxos_command:
        commands: show hostname
        host: "{{ inventory_hostname }}"
        provider: "{{ cli }}"
      register: hostname

    - name: debug hostname
      debug: var=hostname

    - name: db acl
      nxos_acl:
        name: cloud-dbaas-snet-in
        seq: 10
        action: permit
        proto: tcp
        src: any
        dest: any
        state: present
        host: "{{inventory_hostname}}"
        provider: "{{ cli }}"
```

##### EXPECTED RESULTS
I expected a connection to the switch

##### ACTUAL RESULTS

```
~$ ansible-playbook db_acl_test.yml -i lab3k -vvvv
Using /home/robert.nanney/ansible.cfg as config file
Loading callback plugin default of type stdout, v2.0 from /usr/local/lib/python2.7/dist-packages/ansible/plugins/callback/__init__.pyc

PLAYBOOK: db_acl_test.yml ******************************************************
1 plays in db_acl_test.yml

PLAY [all] *********************************************************************

TASK [show hostname] ***********************************************************
task path: /home/robert.nanney/db_acl_test.yml:10
Using module file /usr/local/lib/python2.7/dist-packages/ansible/modules/core/network/nxos/nxos_command.py
<10.127.49.121> ESTABLISH LOCAL CONNECTION FOR USER: robert.nanney
<10.127.49.121> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1479507500.82-139993289090789 `" && echo ansible-tmp-1479507500.82-139993289090789="` echo $HOME/.ansible/tmp/ansible-tmp-1479507500.82-139993289090789 `" ) && sleep 0'
<10.127.49.121> PUT /tmp/tmpffKXca TO /home/robert.nanney/.ansible/tmp/ansible-tmp-1479507500.82-139993289090789/nxos_command.py
<10.127.49.121> EXEC /bin/sh -c 'chmod u+x /home/robert.nanney/.ansible/tmp/ansible-tmp-1479507500.82-139993289090789/ /home/robert.nanney/.ansible/tmp/ansible-tmp-1479507500.82-139993289090789/nxos_command.py && sleep 0'
<10.127.49.121> EXEC /bin/sh -c '/usr/bin/python /home/robert.nanney/.ansible/tmp/ansible-tmp-1479507500.82-139993289090789/nxos_command.py; rm -rf "/home/robert.nanney/.ansible/tmp/ansible-tmp-1479507500.82-139993289090789/" > /dev/null 2>&1 && sleep 0'
ok: [10.127.49.121] => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "auth_pass": null, 
            "authorize": false, 
            "commands": [
                "show hostname"
            ], 
            "host": "10.127.49.121", 
            "interval": 1, 
            "match": "all", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": null, 
            "provider": {
                "transport": "cli"
            }, 
            "retries": 10, 
            "ssh_keyfile": null, 
            "timeout": 10, 
            "transport": "cli", 
            "use_ssl": false, 
            "username": "admin", 
            "validate_certs": true, 
            "wait_for": null
        }, 
        "module_name": "nxos_command"
    }, 
    "stdout": [
        "\nyc3-5-3.sat6 \n"
    ], 
    "stdout_lines": [
        [
            "", 
            "yc3-5-3.sat6 ", 
            ""
        ]
    ], 
    "warnings": []
}

TASK [debug hostname] **********************************************************
task path: /home/robert.nanney/db_acl_test.yml:17
ok: [10.127.49.121] => {
    "hostname": {
        "changed": false, 
        "stdout": [
            "\nyc3-5-3.sat6 \n"
        ], 
        "stdout_lines": [
            [
                "", 
                "yc3-5-3.sat6 ", 
                ""
            ]
        ], 
        "warnings": []
    }
}

TASK [db acl] ******************************************************************
task path: /home/robert.nanney/db_acl_test.yml:20
Using module file /usr/local/lib/python2.7/dist-packages/ansible/modules/core/network/nxos/nxos_acl.py
<10.127.49.121> ESTABLISH LOCAL CONNECTION FOR USER: robert.nanney
<10.127.49.121> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1479507503.1-137909917353841 `" && echo ansible-tmp-1479507503.1-137909917353841="` echo $HOME/.ansible/tmp/ansible-tmp-1479507503.1-137909917353841 `" ) && sleep 0'
<10.127.49.121> PUT /tmp/tmp9S6m5r TO /home/robert.nanney/.ansible/tmp/ansible-tmp-1479507503.1-137909917353841/nxos_acl.py
<10.127.49.121> EXEC /bin/sh -c 'chmod u+x /home/robert.nanney/.ansible/tmp/ansible-tmp-1479507503.1-137909917353841/ /home/robert.nanney/.ansible/tmp/ansible-tmp-1479507503.1-137909917353841/nxos_acl.py && sleep 0'
<10.127.49.121> EXEC /bin/sh -c '/usr/bin/python /home/robert.nanney/.ansible/tmp/ansible-tmp-1479507503.1-137909917353841/nxos_acl.py; rm -rf "/home/robert.nanney/.ansible/tmp/ansible-tmp-1479507503.1-137909917353841/" > /dev/null 2>&1 && sleep 0'
fatal: [10.127.49.121]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "ack": null, 
            "action": "permit", 
            "auth_pass": null, 
            "authorize": false, 
            "config": null, 
            "dest": "any", 
            "dest_port1": null, 
            "dest_port2": null, 
            "dest_port_op": null, 
            "dscp": null, 
            "established": null, 
            "fin": null, 
            "fragments": null, 
            "host": "10.127.49.121", 
            "include_defaults": "False", 
            "log": null, 
            "name": "cloud-dbaas-snet-in", 
            "password": null, 
            "port": null, 
            "precedence": null, 
            "proto": "tcp", 
            "protocol": "http", 
            "provider": {
                "transport": "cli"
            }, 
            "psh": null, 
            "remark": null, 
            "rst": null, 
            "save": false, 
            "seq": "10", 
            "src": "any", 
            "src_port1": null, 
            "src_port2": null, 
            "src_port_op": null, 
            "ssh_keyfile": null, 
            "state": "present", 
            "syn": null, 
            "time_range": null, 
            "timeout": 10, 
            "transport": "cli", 
            "urg": null, 
            "use_ssl": false, 
            "username": null, 
            "validate_certs": true
        }, 
        "module_name": "nxos_acl"
    }, 
    "msg": "failed to connect to 10.127.49.121:22"
}
	to retry, use: --limit @/home/robert.nanney/db_acl_test.retry

PLAY RECAP *********************************************************************
10.127.49.121              : ok=2    changed=0    unreachable=0    failed=1   

```

Copied from original issue: ansible/ansible-modules-core#5675

**Comment by ansibot** [comment]
From @ansibot on 2016-11-18T22:39:18Z



@jedelman8, @GGabriele, @privateip, @gundalow, @qalthos, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @GGabriele on 2016-11-18T22:39:18Z

Can you try adding `username` and `password` to `cli`?

**Comment by ansibot** [comment]
From @RNanney on 2016-11-18T22:39:18Z

@GGabriele: adding username and password to cli worked.

**Comment by ansibot** [comment]
From @GGabriele on 2016-11-18T22:39:18Z

@RNanney awesome! :)

**Comment by ansibot** [comment]
From @RNanney on 2016-11-18T22:39:18Z

Should this module be able to use the ANISIBLE_NET_USERNAME and ANSIBLE_NET_PASSWORD environment variables?

**Comment by ansibot** [comment]
From @GGabriele on 2016-11-18T22:39:18Z

No, these variables are useful when `connection: ssh` to control a remote host.

**Comment by ansibot** [comment]
From @RNanney on 2016-11-18T22:39:18Z

I'm confused because the documentation for the module says the environment variables will be used if username and password are not set. The previous task using the nxos_command module in my example playbook was able to use the environment variables for authentication over ssh.

**Comment by ansibot** [comment]
From @Qalthos on 2016-11-18T22:39:18Z

@GGabriele, unless I am mistaken, this is an issue with nxos_acl, which redefines the `username` and `password` values in its arg_spec dictionary, thus losing the built-in env variable replacement. I think that removing those (and any other accidentally overridden parameters) from the module should restore the functionality.

**Comment by ansibot** [comment]
From @GGabriele on 2016-11-18T22:39:18Z

@Qalthos You're right, I was pretty sure I didn't include them and so couldn't catch them. Opened a PR for this #5851

**Comment by ansibot** [comment]
cc @mikewiebe @rahushen @rcarrillocruz @trishnaguha
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by trishnaguha** [comment]
Hi, can you please retest this on 2.4.0 RC3, we believe it should be fixed.

https://groups.google.com/forum/#!topic/ansible-devel/PjBtOkwbt5I

Thanks
needs_info

**Comment by ansibot** [comment]
cc @kedarX
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@RNanney This issue is waiting for your response. Please respond or the issue will be closed.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@RNanney You have not responded to information requests in this issue so we will assume it no longer affects you. If you are still interested in this, please create a new issue with the requested information.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)
